Fix --env-file flag to work with --dev mode

nsthorat opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/nsthorat/mlflow/pull/17615?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17615/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/17615/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/17615/merge
```

</p>
</details>

## Related Issues/PRs
- Fixes issue where environment variables from `--env-file` are lost during uvicorn auto-reload in development mode

## What changes are proposed in this pull request?

This PR fixes the issue where environment variables loaded via `--env-file` are lost when uvicorn auto-reloads after code changes in development mode.

### Problem
When using `mlflow server --env-file .env --dev`, environment variables are loaded initially but lost when uvicorn auto-reloads. This happens because uvicorn spawns new processes that don't inherit the dynamically loaded environment variables from `dotenv.load_dotenv()`.

### Solution
1. **Pass --env-file to uvicorn**: Pass the `--env-file` path directly to uvicorn's native `--env-file` parameter, ensuring environment variables are properly loaded in each spawned process during reload.

2. **Update envvar for --backend-store-uri**: Updated the option to also check `MLFLOW_TRACKING_URI` environment variable (in addition to `MLFLOW_BACKEND_STORE_URI`) for better compatibility, since `MLFLOW_TRACKING_URI` is the standard environment variable used throughout MLflow.

## How is this PR tested?
- Added `test_server_with_env_file` to verify env_file parameter is passed to uvicorn
- Added `test_build_uvicorn_command_with_env_file` to test uvicorn command building with env_file
- Manually tested server with .env file containing `MLFLOW_TRACKING_URI=databricks` and verified:
  - Connection to Databricks workspace works
  - Environment variables persist after code changes trigger reload
- All existing tests pass

## Does this PR require documentation update?
- No, this is a bug fix that makes the existing `--env-file` flag work correctly with `--dev` mode

## Release Notes
### Is this a user-facing change?
- [x] No. This is only a bug fix with no API change
- [ ] Yes. Make sure to include a milestone for the release of MLflow on which this change will be released.

### What component(s), interfaces, languages, and integrations does this PR affect?
- [x] `area/server-infra`: MLflow Tracking server, tracking client APIs, autologging
- [x] `interface/cli`: Command-line interface for MLflow (mlflow, mlflow run, mlflow server)

### How should the PR be classified in the release notes? Choose one:
- [x] `rn/bug-fix` - A new bug fix

nsthorat committed:
Fix --env-file flag to work with --dev mode

This commit fixes the issue where environment variables from --env-file are
lost during uvicorn's auto-reload in development mode.

The fix:
1. Pass the --env-file path directly to uvicorn's native --env-file parameter,
   ensuring environment variables are properly loaded in each spawned process
   during reload.

2. Update --backend-store-uri to also check MLFLOW_TRACKING_URI environment
   variable for backward compatibility, since MLFLOW_TRACKING_URI is the
   standard environment variable used throughout MLflow.

Signed-off-by: Nikhil Thorat <nikhil.thorat@databricks.com>
Signed-off-by: nsthorat <nsthorat@gmail.com>

BenWilson2 reviewed (approved):
Great quality of life improvement!

nsthorat committed:
Fix test isolation issue in test_server_mlflow_artifacts_options

Add handlers state cleanup similar to other tests to ensure
test runs cleanly in pytest suite.

nsthorat committed:
Remove MLFLOW_TRACKING_URI from backend-store-uri envvar

The addition of MLFLOW_TRACKING_URI to the envvar list for --backend-store-uri
was causing test failures in CI, particularly in artifacts-only mode where
no backend store URI should be set.

MLFLOW_TRACKING_URI and MLFLOW_BACKEND_STORE_URI have different semantics:
- MLFLOW_TRACKING_URI: where to send requests (can be 'databricks', HTTP URLs, etc.)
- MLFLOW_BACKEND_STORE_URI: where to store data (must be storage URIs like sqlite://, file://)

Reverting to only use MLFLOW_BACKEND_STORE_URI for the --backend-store-uri option.

The core env-file functionality still works perfectly.

nsthorat committed:
Merge branch 'master' into env-file-server
